Remove Flux from Signup

jsnajdr opened the pull request:
Remove the progress and dependency Flux stores and dispatch all actions directly to Redux.

All e2e and unit tests pass. Is ready for testing and review.

jsnajdr committed:
Remove Flux from Signup

jsnajdr committed:
Fix reducer for STORE_UPDATE action

jsnajdr committed:
Fixup proptypes in SiteTopic

jsnajdr committed:
Remove the concept of unstorableDependencies, doesn't work

jsnajdr committed:
Fix the step components' tests

jsnajdr committed:
Fix tests of dependency store reducer

jsnajdr committed:
Fixup the dependency store reducer tests

jsnajdr committed:
Fixup the progress store reducer tests

jsnajdr committed:
Fixup flow controller tests

jsnajdr committed:
Fixup step actions unit tests

scruffian commented:
Thanks for all your work on this. I've read through the code and the changes all make sense to me. I tested the main flow and it all seems to work as expected.

jsnmoon commented on client/signup/steps/reader-landing/index.jsx:
```diff
@@ -20,7 +22,8 @@ import './style.scss';
 class ReaderLandingStep extends Component {
 	handleButtonClick = () => {
-		SignupActions.submitSignupStep( { stepName: this.props.stepName } );
+		this.props.recordTracksEvent( 'calypso_signup_reader_landing_cta' );
```
Was this an intentional addition?

jsnajdr replied:
Yes, here I merged the two `handleButtonClick` handlers in `ReaderLandingStep` and `ReacerLandingStepContent` into one. The inner component now just passes down a handler prop and doesn't need to be connected to Redux.

jsnmoon commented on client/signup/steps/site-style/index.jsx:
```diff
@@ -58,13 +55,20 @@ export class SiteStyleStep extends Component {
+		this.submitSiteStyle( selectedStyleData.id, selectedStyleData.theme, selectedStyleData.label );
 	};
 
+	submitSiteStyle( siteStyle, themeSlugWithRepo, styleLabel ) {
```
Nice!

jsnmoon commented on client/lib/signup/flow-controller.js:
```diff
@@ -4,36 +4,50 @@
  */
```
This is a beast to review. Everything seemed okay on my first pass, but I'm still a bit wary of this module.

jsnmoon reviewed (commented):
Great work, @jsnajdr. It's really nice to see this finally coming to fruition.

When are we removing flux from Calypso entirely? 😏

niranjan-uma-shankar commented:
Just dropping in to say this is truly an awesome effort. Kudos @jsnajdr!

adamziel commented:
I have no words, it's just incredible what you did here - me and everyone else working on signup are in debt (although not technical debt anymore!) @jsnajdr - thank you!

spen reviewed (approved):
Awesome! Thanks heaps for this @jsnajdr - it can't have been easy!
I didn't see anything that stood out while as an issue while reading through the code - it's a tough one to review with 100% confidence just because of the weight of the changes - I'd definitely be putting some faith in to your work here, as opposed to understanding 100% of all the changes, so that's worth baring in mind - we might want to get a few ✅s from others before merging, and making sure we have a few people around when we do merge so we can test it well while in staging.  
I've given it a few run throughs too and found no hitches!

jsnajdr commented on client/lib/signup/step-actions.js:
```diff
@@ -182,7 +182,10 @@ export function createSiteWithCart(
 	) {
-		newSiteParams.blog_name = get( user.get(), 'username', siteTitle ) || siteType || siteVertical;
+		const username =
+			getCurrentUserName( state ) || dependencies.username || siteTitle || siteType || siteVertical;
```
It's the username entered in the `user` step. It's in the dependency store.

I moved this patch to a separate PR (#33309) as it's not really relevant to Flux. I only discovered the bug when working on the Flux removal.

I had to modify the `dependencies.username` code a bit in #33309 as the `dependencies` variable contains dependencies only for the step that calls the `createSiteWithCart` API function, not all of them.

jsnajdr commented on client/lib/signup/step-actions.js:
```diff
@@ -182,7 +182,10 @@ export function createSiteWithCart(
+		const username =
+			getCurrentUserName( state ) || dependencies.username || siteTitle || siteType || siteVertical;
+
+		newSiteParams.blog_name = username;
```
Yes, I implemented this suggestion in #33309 🙂 Thanks!

jsnajdr commented:
Thanks everyone for your nice comments! I'm glad you like it. I rebased the PR onto the latest master to resolve some conflicts. And moved some changes to separate PRs: #33306, #33309 and #33311. It makes this diff smaller, although admittedly still difficult to review.

I'll go through the review comments on Monday and will address them :+1:

jsnajdr commented on client/state/signup/progress/test/reducer.js:
```diff
@@ -29,21 +28,110 @@ jest.mock( 'signup/config/steps-pure', () => ( {
+			step: { stepName: 'site-selection' },
+		} );
+
+		state = reducer( state, {
```
That's a good suggestion, I'll implement it. I realized that the `dispatchSequence` function is already part of JavaScript:
```js
const finalState = actions.reduce( reducer, initialState );
```
:smile:

jsnajdr replied:
Suggestion implemented in 0a53f1a.

spen replied:
😍Awesome, and elegant as always buddy!
I'll be stealing this for sure! ;)

jsnajdr committed:
Signup Progress: improve test code by smart use of Array.reduce

jsnajdr committed:
Passwordless Step: migrate to Redux

jsnajdr commented:
6042b13 migrates the `PasswordlessStep` component (recently added by @scruffian) that was using the old-style Flux actions.

jsnajdr commented on client/lib/signup/step-actions.js:
```diff
@@ -452,10 +455,22 @@ export function createAccount(
+
+				if ( ! response || ! response.bearer_token ) {
+					// something odd happened...
+					//eslint-disable-next-line no-console
```
Only if we removed the `console.error` call. And we want to keep it. I see the `no-console` rule mostly as a safeguard against forgotten debug printouts. That's not the case here.

jsnajdr commented on client/lib/signup/step-actions.js:
```diff
@@ -487,32 +502,34 @@ export function createAccount(
 					error && error.error ? [ { error: error.error, message: error.message } ] : undefined;
+
+				if ( errors ) {
+					callback( errors );
```
Yes, it's ok because in case of error, the `providedDependencies` param would go unused anyway. The callback dispatches the `processedSignupStep` action (renamed to `completeSignupStep`). If the action has the `errors` property, then [`invalidateStep` is performed and the rest is skipped](https://github.com/Automattic/wp-calypso/blob/ef6f2ed286d42f3a7b1fbe674317751acaa9641a/client/lib/signup/progress-store.js#L108-L111).

jsnajdr committed:
Update the SiteStyle button test snapshot
